### DISCUSSION TITLE
The order of the detected words is not accurate.

### DISCUSSION THREAD

**Discussion by vdt104** [mindee/doctr#1475] [Q&A]
I use doctr for text detection. My images are usually photos of documents like the one below. Doctr performs very well in text detection, but the order of the words is not accurate.

![image](https://github.com/mindee/doctr/assets/104601534/0c0469d5-1b5b-41b2-ad5b-b3f30127891e)

Ex: 
doctr detect:
((0.2575127110514198, 0.009765625), (0.432246198676132, 0.0419921875)): ERNEST
((0.4375951829911742, 0.0), (0.7175253621450499, 0.0283203125)): HEMINGWAY
((0.2860406273983116, 0.06640625), (0.38232234506907137, 0.0966796875)): này,
((0.3787563555257099, 0.0654296875), (0.43402919344781277, 0.0859375)): xa
((0.432246198676132, 0.0576171875), (0.49643401045663854, 0.0830078125)): bờ
((0.49643401045663854, 0.052734375), (0.5838007542689946, 0.080078125)): đến
((0.5873667438123561, 0.0517578125), (0.6747334876247122, 0.080078125)): thế,
((0.6782994771680737, 0.0517578125), (0.771015205295472, 0.0791015625)): hẳn
((0.771015205295472, 0.0595703125), (0.8387690066193401, 0.0830078125)): nó
((0.8405520013910207, 0.060546875), (0.8940418445414429, 0.087890625)): là
((0.10239216591519568, 0.08203125), (0.18797591495587107, 0.109375)): Vào
((0.1790609410974674, 0.076171875), (0.2931726064850345, 0.10546875)): tháng
.....

label: ERNEST HEMINGWAY, vào tháng này, xa bờ đến thế, hẳn nó là con cá ....

Is there a solution to this problem?

**Comment by felixdittrich92** [comment]
Hi @104-wonohfor 👋,

The result from 
```
res = predictor(doc)
res_text = res.render()
```
Is also not correctly sorted ?

**Reply by felixdittrich92** [reply]
If you use only the detection predictor the sorting is in user hands atm. 
Because sorting is performed only with the whole Pipeline (DocumentBuilder) :)

**Reply by vdt104** [reply]
The result is also not correctly sorted. Here is code:

```
from doctr.io import DocumentFile
from doctr.models import ocr_predictor

model = ocr_predictor(det_arch = 'db_resnet50', reco_arch = 'crnn_mobilenet_v3_large',pretrained=True)

img_path = ""
doc = DocumentFile.from_images(img_path)

res = model(doc)
res_text = res.render()
res_text
```
Output: 
'ERNEST HEMINGWAY\n\nnày, xa bo den the, hân no là\n\nVào thâng nghi. Dop mdi di, câ.

Only some first lines (not straight like image above) are not correctly sorted. Other lines which are at the middle and the end are accurate.

**Reply by felixdittrich92** [reply]
Mh yeah i see i have had already in mind to add something optional for unwarping as a preprocessing step.(Like the doctr++ model)
How does it look If you use (`assume_straight_pages=False` to the ocr_predictor instance - only main branch - v0.8.0a) ?

**Reply by vdt104** [reply]
I have tried, the result is still the same. 

model = ocr_predictor('db_resnet50_rotation', pretrained=True, assume_straight_pages=False, preserve_aspect_ratio=True)

**Reply by vdt104** [reply]
I think this problem may due to the detect line process.  
![image](https://github.com/mindee/doctr/assets/104601534/2519e78f-822d-43ec-aae6-b33617008fc1)
The bounding box is retrieved from output[lines]

**Reply by felixdittrich92** [reply]
Yes, the problem is the text curvature, unfortunately I don't think we can do much about that at the moment. 
However, as already written, it is planned to offer a solution for pre-processing in the future.
